### ISSUE TITLE
[🐛 Bug]: Selenium4: Handled exception stacktraces in log output

### ISSUE THREAD

**Issue by leRobbe** [SeleniumHQ/selenium#9974]
### What happened?

I'm currently trying to upgrade from Selenium 3.141.59 to 4.0.0 and I've encountered some issues, one of which is described in the following.

I try to find an element where I know it might not exist depending on the previous steps in the application, so I surrounded it with a try/catch block where I catch _NoSuchElementException_ as well as _TimeoutException_ since I use a _FluentWait_ to poll for a specified amount of time. When the element was found everything is fine. However, if it's not found, the exception handling occurs as intended and the test progresses accordingly, BUT the stacktraces of both aforementioned exceptions are still written to the log, even though I never specifically output any of them myself. In the exception handling I just say catch exception xyz and do nothing.

I've also discovered that stacktraces showing in logs only occurs when using the listener we've implemented that waits for an element to be present etc before .findBy(..) is executed, since the application can be a tad bit slow at times (especially when debugging locally). When I don't use the listener but copy its code to the test code itself, everything works as expected..exception handling occurs and no stacktraces of exceptions that were caught show up on the log.

### How can we reproduce the issue?

DriverEventListener:
```
public class DriverEventListener implements WebDriverListener {
    private void waitForElement(WebDriver driver, By selector) {

        Wait<WebDriver> wait = new FluentWait<>(driver).withTimeout(Duration.ofSeconds(4))
            .pollingEvery(Duration.ofSeconds(1)).ignoring(NoSuchElementException.class);

        if (selector != null) {
            wait.until(ExpectedConditions.presenceOfElementLocated(selector));
            wait.until(driver1 -> {
                WebElement element = driver1.findElement(selector);
                return (element != null && element.isEnabled() ? element : null);
            });
        }
    }
    
    @Override
    public void beforeFindElement(WebDriver driver, By locator) {
        waitForElement(driver, locator);
    }
}
```

The following code works fine (no exception stacktraces, NO listener used):
```
public class ThisWorksFine {
    public static void main(String[] args) {
        System.setProperty("webdriver.gecko.driver", "D:/Workspaces/selenium/geckodriver-v0.30.0/geckodriver.exe");
        WebDriver driver = new FirefoxDriver();

        driver.get("https://google.com");

        Wait<WebDriver> wait = new FluentWait<>(driver).withTimeout(Duration.ofSeconds(4))
            .pollingEvery(Duration.ofSeconds(1)).ignoring(NoSuchElementException.class);

        try {
            By selector = By.id("username"); // this ID doesn't exist on given page
            wait.until(ExpectedConditions.presenceOfElementLocated(selector));
            wait.until(driver1 -> {
                WebElement element = driver1.findElement(selector);
                return (element != null && element.isEnabled() ? element : null);
            });
        } catch (NoSuchElementException | TimeoutException ex) {
            System.out.println("Some exception handling happens");
        } finally {
            driver.quit();
        }
    }
}
```

With output:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
1635408507388	geckodriver	INFO	Listening on 127.0.0.1:52355
1635408508273	mozrunner::runner	INFO	Running command: "C:\\Program Files (x86)\\Mozilla Firefox\\firefox.exe" "--marionette" "--remote-debugging-port" "51536" "-no-remote" "-profile" "D:\\Users\\mmg\\AppData\\Local\\Temp\\rust_mozprofilekg9ybW"
1635408508646	Marionette	INFO	Marionette enabled
console.warn: SearchSettings: "get: No settings file exists, new profile?" (new NotFoundError("Could not open the file at D:\\Users\\removed\\AppData\\Local\\Temp\\rust_mozprofilekg9ybW\\search.json.mozlz4", (void 0)))
DevTools listening on ws://localhost:51536/devtools/browser/cf2ed46a-d572-401e-a3c8-c761f24b79ba
1635408510020	Marionette	INFO	Listening on port 54010
1635408510406	RemoteAgent	WARN	TLS certificate errors will be ignored for this session
Okt 28, 2021 10:08:30 AM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
Some exception handling happens
1635408516586	Marionette	INFO	Stopped listening on port 54010
JavaScript error: resource:///modules/Interactions.jsm, line 201: NS_ERROR_FAILURE: Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIUserIdleService.removeIdleObserver]
```

This DOESN'T work (but is the code we need to work; listener used):
```
public class ThisDoesntWork {
    public static void main(String[] args) {
        System.setProperty("webdriver.gecko.driver", "D:/Workspaces/selenium/geckodriver-v0.30.0/geckodriver.exe");
        FirefoxDriver temp = new FirefoxDriver();
        WebDriver driver = new EventFiringDecorator(new DriverEventListener()).decorate(temp);

        driver.get("https://google.com");

        try {
            By selector = By.id("username"); // this ID doesn't exist on given page
            driver.findElement(selector);
        } catch (NoSuchElementException | TimeoutException ex) {
            System.out.println("Some exception handling happens");
        } finally {
            driver.quit();
        }
    }
}
```

With output:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
1635408622214	geckodriver	INFO	Listening on 127.0.0.1:53561
1635408623658	mozrunner::runner	INFO	Running command: "C:\\Program Files (x86)\\Mozilla Firefox\\firefox.exe" "--marionette" "--remote-debugging-port" "49191" "-no-remote" "-profile" "D:\\Users\\mmg\\AppData\\Local\\Temp\\rust_mozprofile2i2KJm"
1635408624064	Marionette	INFO	Marionette enabled
console.warn: SearchSettings: "get: No settings file exists, new profile?" (new NotFoundError("Could not open the file at D:\\Users\\mmg\\AppData\\Local\\Temp\\rust_mozprofile2i2KJm\\search.json.mozlz4", (void 0)))
DevTools listening on ws://localhost:49191/devtools/browser/36f17087-29a9-4bd9-8225-46520f5a6dd8
1635408625713	Marionette	INFO	Listening on port 54073
1635408625942	RemoteAgent	WARN	TLS certificate errors will be ignored for this session
Okt 28, 2021 10:10:26 AM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
Okt 28, 2021 10:10:31 AM org.openqa.selenium.support.events.EventFiringDecorator callListenerMethod
WARNUNG: null
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:90)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
	at java.lang.reflect.Method.invoke(Method.java:508)
	at org.openqa.selenium.support.events.EventFiringDecorator.callListenerMethod(EventFiringDecorator.java:311)
	at org.openqa.selenium.support.events.EventFiringDecorator.fireBeforeEvents(EventFiringDecorator.java:231)
	at org.openqa.selenium.support.events.EventFiringDecorator.lambda$beforeCall$0(EventFiringDecorator.java:170)
	at org.openqa.selenium.support.events.EventFiringDecorator$$Lambda$304.0000000012238030.accept(Unknown Source)
	at java.util.Arrays$ArrayList.forEach(Arrays.java:4224)
	at org.openqa.selenium.support.events.EventFiringDecorator.beforeCall(EventFiringDecorator.java:170)
	at org.openqa.selenium.support.decorators.DefaultDecorated.beforeCall(DefaultDecorated.java:43)
	at org.openqa.selenium.support.decorators.WebDriverDecorator.lambda$createProxy$0(WebDriverDecorator.java:293)
	at org.openqa.selenium.support.decorators.WebDriverDecorator$$Lambda$303.0000000012237A50.invoke(Unknown Source)
	at com.sun.proxy.$Proxy12.findElement(Unknown Source)
	at DeleteThisToo.main(ThisDoesntWork.java:20)
Caused by: org.openqa.selenium.TimeoutException: Expected condition failed: waiting for presence of element located by: By.id: username (tried for 4 second(s) with 1000 milliseconds interval)
	at org.openqa.selenium.support.ui.FluentWait.timeoutException(FluentWait.java:263)
	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:231)
	at de.base.DriverEventListener.waitForElement(DriverEventListener.java:54)
	at de.base.DriverEventListener.beforeFindElement(DriverEventListener.java:142)
	... 15 more
Caused by: org.openqa.selenium.NoSuchElementException: Unable to locate element: #username
For documentation on this error, please visit: https://selenium.dev/exceptions/#no_such_element
Build info: version: '4.0.0', revision: '3a21814679'
System info: host: 'EBA-VM245', ip: '192.168.203.181', os.name: 'Windows 8.1', os.arch: 'amd64', os.version: '6.3', java.version: '1.8.0_151'
Driver info: org.openqa.selenium.firefox.FirefoxDriver
Command: [730222db-a063-4dd3-abc6-3711bc68ff76, findElement {using=id, value=username}]
Capabilities {acceptInsecureCerts: true, browserName: firefox, browserVersion: 93.0, javascriptEnabled: true, moz:accessibilityChecks: false, moz:buildID: 20210927210923, moz:debuggerAddress: localhost:49191, moz:geckodriverVersion: 0.30.0, moz:headless: false, moz:processID: 7724, moz:profile: D:\Users\mmg\AppData\Local\..., moz:shutdownTimeout: 60000, moz:useNonSpecCompliantPointerOrigin: false, moz:webdriverClick: true, pageLoadStrategy: normal, platform: WINDOWS, platformName: WINDOWS, platformVersion: 6.3, proxy: Proxy(), se:cdp: ws://localhost:49191/devtoo..., se:cdpVersion: 85, setWindowRect: true, strictFileInteractability: false, timeouts: {implicit: 0, pageLoad: 300000, script: 30000}, unhandledPromptBehavior: dismiss and notify}
Session ID: 730222db-a063-4dd3-abc6-3711bc68ff76
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:83)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:57)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:437)
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.createException(W3CHttpResponseCodec.java:200)
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:133)
	at org.openqa.selenium.remote.codec.w3c.W3CHttpResponseCodec.decode(W3CHttpResponseCodec.java:53)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:184)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.invokeExecute(DriverCommandExecutor.java:164)
	at org.openqa.selenium.remote.service.DriverCommandExecutor.execute(DriverCommandExecutor.java:139)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:559)
	at org.openqa.selenium.remote.ElementLocation$ElementFinder$2.findElement(ElementLocation.java:162)
	at org.openqa.selenium.remote.ElementLocation.findElement(ElementLocation.java:60)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:383)
	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:375)
	at org.openqa.selenium.support.ui.ExpectedConditions$6.apply(ExpectedConditions.java:174)
	at org.openqa.selenium.support.ui.ExpectedConditions$6.apply(ExpectedConditions.java:171)
	at org.openqa.selenium.support.ui.FluentWait.until(FluentWait.java:208)
	... 17 more

Some exception handling happens
1635408631281	Marionette	INFO	Stopped listening on port 54073
JavaScript error: resource:///modules/Interactions.jsm, line 201: NS_ERROR_FAILURE: Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIUserIdleService.removeIdleObserver]
```

(Note that the JavaScript error in both cases started occurring upon upgrading from Selenium 3.141.59 to 4.0.0)

As you can see, in the given example above where we use the listener, exception handling occurs as expected, but the stacktraces of the caught exceptions still show up in the logs. 

If the behavior described above (stacktraces printed to logs even when exception is handled) is the intended behavior for the setup with the listener indeed, is there any way to suppress log outputs in case the exception is handled (but ofc output it if an unintended TimeoutException etc occured at some other point)? Also, why do stacktraces occur in the setup with the listener whereas everything is as expected when not using the listener but identical code? Anything I'm doing wrong here that I might be missign?

### What browsers and operating systems are you seeing the problem on?

Browser: Firefox
Geckodriver version 0.30.0
OS: Windows 8.1
Selenium 4.0.0 (no grid)

**Comment by leRobbe** [comment]
P.S. Sorry I think this issue somehow landed in the wrong category, don't know what I did wrong for it not to land in the general issues section..

**Comment by diemol** [comment]
This issue was created in the docs repo and then transferred here. Information for triaging this is missing. Please either consider creating a new one with all the information, or check the [template](https://github.com/SeleniumHQ/selenium/issues/new?assignees=&labels=I-defect%2Cneeds-triaging&template=bug-report.yml&title=%5B%F0%9F%90%9B+Bug%5D%3A+) and add the information to the text above.

**Comment by leRobbe** [comment]
Sorry for the late response.
I've updated the initial issue description with code examples to reproduce and according log outputs.

If there's any other information that you need me to provide please let me know.

**Comment by leRobbe** [comment]
Has anyone had a chance to look at this issue yet?
As a temporary solution it would also work if I could somehow disable the driver log output, but I haven't found a way to do so yet.

**Comment by pujagani** [comment]
Thank you for sharing the details and the reproducible test script. I was able to reproduce the issue. However, after debugging and stepping through the code, the behavior when an error occurs and the error handler is called is the same in both cases (with and without the listener). When `driver.findElement()` is called it is expected to throw an `NoSuchElementException` exception. Since it uses until `TimeoutException` is excepted as well. This holds true when it is used in any part of the code including the method called from `beforeFindElement` method of the listener. It is the intended behavior.

Hence, the output shared has the exception log that is coming from the ` WebElement element = driver1.findElement(selector);` of the `DriverEventListener` class. Surrounding that portion of the code with try/catch should do the trick. 

```
public class DriverEventListener implements WebDriverListener {
  private void waitForElement(WebDriver driver, By selector) {

    try {
      Wait<WebDriver> wait = new FluentWait<>(driver).withTimeout(Duration.ofSeconds(4))
        .pollingEvery(Duration.ofSeconds(1)).ignoring(NoSuchElementException.class);

      if (selector != null) {
        wait.until(ExpectedConditions.presenceOfElementLocated(selector));
        wait.until(driver1 -> {
          WebElement element = driver1.findElement(selector);
          return (element != null && element.isEnabled() ? element : null);
        });
      }
    } catch (NoSuchElementException | TimeoutException ex) {
      System.out.println("Some exception handling happens in DriverEventListener");
    }
  }

  @Override
  public void beforeFindElement(WebDriver driver, By locator) {
    waitForElement(driver, locator);
  }
}
```

Output:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
1652852363409	geckodriver	INFO	Listening on 127.0.0.1:60988
1652852363763	mozrunner::runner	INFO	Running command: "/Applications/Firefox.app/Contents/MacOS/firefox-bin" "--marionette" "--remote-debugging-port" "60216" "--remote-allow-hosts" "localhost" "-foreground" "-no-remote" "-profile" "/var/folders/gf/wjxc6bv113b08xm754y0x0_h0000gp/T/rust_mozprofileFfrJNd"
1652852364310	Marionette	INFO	Marionette enabled
WebDriver BiDi listening on ws://localhost:60216
2022-05-18 11:09:24.982 firefox-bin[2837:66041] Warning: Expected min height of view: (<NSButton: 0x137723000>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2022-05-18 11:09:24.982 firefox-bin[2837:66041] Warning: Expected min height of view: (<NSButton: 0x137727800>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2022-05-18 11:09:24.983 firefox-bin[2837:66041] Warning: Expected min height of view: (<NSButton: 0x137729c00>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2022-05-18 11:09:24.983 firefox-bin[2837:66041] Warning: Expected min height of view: (<NSButton: 0x137729800>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2022-05-18 11:09:24.984 firefox-bin[2837:66041] Warning: Expected min height of view: (<NSButton: 0x137724800>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2022-05-18 11:09:24.984 firefox-bin[2837:66041] Warning: Expected min height of view: (<NSPopoverTouchBarItemButton: 0x137725400>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
2022-05-18 11:09:24.985 firefox-bin[2837:66041] Warning: Expected min height of view: (<NSPopoverTouchBarItemButton: 0x137730000>) to be less than or equal to 30 but got a height of 32.000000. This error will be logged once per view in violation.
console.warn: SearchSettings: "get: No settings file exists, new profile?" (new NotFoundError("Could not open the file at /var/folders/gf/wjxc6bv113b08xm754y0x0_h0000gp/T/rust_mozprofileFfrJNd/search.json.mozlz4", (void 0)))
console.error: BackgroundUpdate: 
  _reasonsToNotScheduleUpdates: Failed to check for Maintenance Service Registry Key: [Exception... "Component returned failure code: 0x80004001 (NS_ERROR_NOT_IMPLEMENTED) [nsIUpdateProcessor.getServiceRegKeyExists]"  nsresult: "0x80004001 (NS_ERROR_NOT_IMPLEMENTED)"  location: "JS frame :: resource://gre/modules/BackgroundUpdate.jsm :: _reasonsToNotScheduleUpdates :: line 241"  data: no]
1652852365912	Marionette	INFO	Listening on port 52696
DevTools listening on ws://localhost:60216/devtools/browser/4ec8fe7d-ead9-42ab-a07c-1fddfa6cbf94
Read port: 52696
1652852366023	RemoteAgent	WARN	TLS certificate errors will be ignored for this session
May 18, 2022 11:09:26 AM org.openqa.selenium.remote.ProtocolHandshake createSession
INFO: Detected dialect: W3C
Some exception handling happens in DriverEventListener
Some exception handling happens
1652852373437	Marionette	INFO	Stopped listening on port 52696

###!!! [Parent][PImageBridgeParent] Error: RunMessage(msgname=PImageBridge::Msg_WillClose) Channel closing: too late to send/recv, messages will be lost
```

Please try out the above-suggested solution. Thank you!

**Comment by leRobbe** [comment]
Ok thank you, I will try your solution as soon as I have time (kinda busy schedule atm)!

However, I'm still not quite sure I understand why with Selenium4 there's this "extra" error log output now, which did not occur with Selenium 3.141.59 (unless that was a "lucky bug" we had with the previous version).

**Comment by pujagani** [comment]
The way WebDriverListener was implemented in 3.141.59 and in Selenium 4 is very different. Though the core feature remains intact, the underlying implementation difference could be one potential reason. 

Please provide feedback when time permits. We can reopen the issue as needed.

**Comment by leRobbe** [comment]
Ah I see, that probably explains a lot^^;
I hope I'll get to it within the next 2 weeks or so and will let you know how things worked out for us =)
